### PR TITLE
Next/Previous Bookmark Gestures

### LINKED ISSUE #5965: Gestures for jumping to next or previous bookmark?
Would it be possible to have menu items for jumping to the next or previous bookmark?

I know this is possible with the skimmer, but the skimmer when active covers almost a third of the visible page, so it is not really handy for jumping from bookmark to bookmark. To read the page I skimmed to I have to close the skimmer, then re-invoke it to jump to the next bookmark, close the skimmer, read the page, re-invoke the skimmer, etc. Tedious!

When gestures could be added for jumping back and forth between bookmarks (just like double tapping for jumping between chapters), this would be a real power feature!

### PR DESCRIPTION
**Pull request by yparitcher**
Fixes #5965

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/5968)

### CODE REVIEW

File: frontend/apps/reader/modules/readerbookmark.lua [hunk 1]
```diff
@@ -209,8 +209,10 @@ function ReaderBookmark:onPosUpdate(pos)
 end
 
 function ReaderBookmark:gotoBookmark(pn_or_xp)
-    local event = self.ui.document.info.has_pages and "GotoPage" or "GotoXPointer"
-    self.ui:handleEvent(Event:new(event, pn_or_xp))
+    if pn_or_xp then
```
Comment (Frenzie) [inline frontend/apps/reader/modules/readerbookmark.lua]: Is this related to a specific problem you came across?

File: frontend/apps/reader/modules/readergesture.lua [hunk 2]
```diff
@@ -1631,4 +1639,14 @@ function ReaderGesture:onToggleReadingOrder()
     return true
 end
 
+function ReaderGesture:getCurrentPage()
```
Comment (Frenzie) [inline frontend/apps/reader/modules/readergesture.lua]: We don't already have something like this? And are we sure this is where it should go?

File: frontend/apps/reader/modules/readerbookmark.lua (hunk 1 above)
Comment (yparitcher) [reply frontend/apps/reader/modules/readerbookmark.lua]: at the first bookmark and previous, or last bookmark and next.
`pn_or_xp` will be nill crashing koreader.
instead of putting a check in every function, i put one check at execution

File: frontend/apps/reader/modules/readergesture.lua (hunk 2 above)
Comment (yparitcher) [reply frontend/apps/reader/modules/readergesture.lua]: not sure, if we do please let me know it would be helpfull all over the place :)
right now we have a bunch of 
```lua
    if self.document.info.has_pages then
        page = self.ui.paging.current_page
    else
        page = self.document:getCurrentPage()
    end
```
all over the codebase. so i assumed there wasnt one yet.

In general i think this should be fixed everywhere

Comment (Frenzie) [reply frontend/apps/reader/modules/readergesture.lua]: I think something like `Document:getCurrentPage` would probably make the most sense.

File: frontend/apps/reader/modules/readergesture.lua [hunk 3]
```diff
@@ -1300,6 +1304,10 @@ function ReaderGesture:gestureAction(action, ges)
         self.ui:handleEvent(Event:new("GotoNextChapter"))
     elseif action == "prev_chapter" then
         self.ui:handleEvent(Event:new("GotoPrevChapter"))
+    elseif action == "next_bookmark" then
+        self.ui:handleEvent(Event:new("GotoNextBookmarkFromPage", self:getCurrentPage()))
```
Comment (poire-z) [inline frontend/apps/reader/modules/readergesture.lua]: Does this getCurrentPage() business have to really be in this readergesture file?
Can't we have just `self.ui:handleEvent(Event:new("GotoNextBookmarkFromCurrentPage"))` and have the getting of the current page in the 2 functions you added?

File: frontend/apps/reader/modules/readergesture.lua (hunk 2 above)
Comment (yparitcher) [reply frontend/apps/reader/modules/readergesture.lua]: a qiuck `ag` shows
https://github.com/koreader/koreader/blob/8d8b8cd0f79455ac4e4bc1dc25524a98a24cdeea/frontend/apps/reader/modules/readergesture.lua#L1568
https://github.com/koreader/koreader/blob/8d8b8cd0f79455ac4e4bc1dc25524a98a24cdeea/frontend/apps/reader/skimtowidget.lua#L66
https://github.com/koreader/koreader/blob/8d8b8cd0f79455ac4e4bc1dc25524a98a24cdeea/frontend/apps/reader/skimtowidget.lua#L424
https://github.com/koreader/koreader/blob/8d8b8cd0f79455ac4e4bc1dc25524a98a24cdeea/plugins/statistics.koplugin/main.lua#L680

**Comment by yparitcher** [comment]
> I think something like `Document:getCurrentPage()` would probably make the most sense.

CreDocument already implements this.
the other Mupdf based ones don't, as the pages are managed in `ReaderPaging`. what would be the best way to get the `paging.current_page` value in `Document:`?

**Comment by poire-z** [comment]
```lua
    if self.document.info.has_pages then
        curr_page = self.ui.paging.current_page
    else
        curr_page = self.document:getCurrentPage()
    end
```

> what would be the best way to get the paging.current_page value in Document:?

If you're asking, may be there's none obvious :) and that's why it's the way it is.
May be forcing it to be a method of Document is not the most logical way (haven't checked), even if it looks like it would in the snippets you mentionned.

(I'd say: don't bother "fixing" that for such a little PR :)

**Comment by yparitcher** [comment]
I think i managed it. Let me know if this looks good or i should abandon the whole mess

File: frontend/apps/reader/readerui.lua [hunk 4]
```diff
@@ -270,6 +270,10 @@ function ReaderUI:init()
             ui = self,
             document = self.document,
         })
+        -- Current page
+        self.getCurrentPage = function()
+          return self.paging.current_page
+        end
```
Comment (poire-z) [inline frontend/apps/reader/readerui.lua]: For clarity, I think these could just be regular methods: `ReaderUI:getCurrentPage()` with a `if self.document.info.has_pages then` in it (like we have everywhere).
Just so we find it where we expect to find it, instead of having to think about a dynamically plugged function.

(And you call it like a methods, while it really is an unbound function.)

Comment (yparitcher) [reply frontend/apps/reader/readerui.lua]: done

Comment (yparitcher) [reply frontend/apps/reader/readerui.lua]: what is the difference between lua methods and functions?

Comment (poire-z) [reply frontend/apps/reader/readerui.lua]:
```
Object.function(arg1, arg2)
Object:method(arg1, arg2)
```
The `:` implicitely defines self as being expected as the 1st argument, and when you call `instance:method(arg1, arg2)`, it implicitely passes `instance` as `self` in the call.

In the definitions, these are equivalents, but the 1st is more readable:
```
Object:method(arg1, arg2)
Object.function(self, arg1, arg2)
```

Comment (yparitcher) [reply frontend/apps/reader/readerui.lua]: thanks for that, lua is not my primary language :)

File: frontend/apps/reader/readerui.lua [hunk 5]
```diff
@@ -747,4 +747,12 @@ function ReaderUI:switchDocument(new_file)
     self:showReader(new_file)
 end
 
+function ReaderUI:getCurrentPage()
```
Comment (poire-z) [inline frontend/apps/reader/readerui.lua]: Was about to suggest to put it as a more appropriate place (that at the end), with some other alike functions... but there are none :)
So, indeed, it's not the module where we expect such functions... so let it be there.
May be just add a comment `-- little convenient helper`
(Or move if into ReaderView, where it would be more at home?)

File: frontend/apps/reader/modules/readergesture.lua [hunk 6]
```diff
@@ -1300,6 +1304,10 @@ function ReaderGesture:gestureAction(action, ges)
         self.ui:handleEvent(Event:new("GotoNextChapter"))
     elseif action == "prev_chapter" then
         self.ui:handleEvent(Event:new("GotoPrevChapter"))
+    elseif action == "next_bookmark" then
+        self.ui:handleEvent(Event:new("GotoNextBookmarkFromPage", self.ui:getCurrentPage()))
+    elseif action == "prev_bookmark" then
+        self.ui:handleEvent(Event:new("GotoPreviousBookmarkFromPage", self.ui:getCurrentPage()))
```
Comment (poire-z) [inline frontend/apps/reader/modules/readergesture.lua]: Still want to pass current page from there?

Comment (yparitcher) [reply frontend/apps/reader/modules/readergesture.lua]: I dont care either way, i am just keeping to the style in `readerbookmark` where all the functions are passed a page.

Comment (poire-z) [reply frontend/apps/reader/modules/readergesture.lua]: I prefer as I proposed in https://github.com/koreader/koreader/pull/5968#discussion_r392711103 (no need to pass a page if the function/event is named ...FromCurrentPage)

File: frontend/apps/reader/readerui.lua [hunk 7]
```diff
@@ -320,6 +324,11 @@ function ReaderUI:init()
             ui = self
         })
         self.disable_double_tap = G_reader_settings:readSetting("disable_double_tap") ~= false
+        -- Current page
+        self.getCurrentPage = function()
+          return self.document:getCurrentPage()
```
Comment (Frenzie) [inline frontend/apps/reader/readerui.lua]: nit

File: frontend/apps/reader/readerui.lua [hunk 8]
```diff
@@ -270,6 +270,10 @@ function ReaderUI:init()
             ui = self,
             document = self.document,
         })
+        -- Current page
+        self.getCurrentPage = function()
+          return self.paging.current_page
```
Comment (Frenzie) [inline frontend/apps/reader/readerui.lua]: nit

File: frontend/apps/reader/modules/readergesture.lua (hunk 6 above)
Comment (yparitcher) [reply frontend/apps/reader/modules/readergesture.lua]: done

File: frontend/apps/reader/skimtowidget.lua [hunk 9]
```diff
@@ -421,11 +395,7 @@ function SkimToWidget:goToBookmark(page)
     if page then
         self:addOriginToLocationStack()
         self.ui.bookmark:gotoBookmark(page)
-        if self.document.info.has_pages then
-            self.curr_page = self.ui.paging.current_page
-        else
-            self.curr_page = self.document:getCurrentPage()
-        end
+        self.curr_page  = self.ui:getCurrentPage()
```
Comment (Frenzie) [inline frontend/apps/reader/skimtowidget.lua]: more nit :-P

```suggestion
        self.curr_page = self.ui:getCurrentPage()
```

File: plugins/statistics.koplugin/main.lua [hunk 10]
```diff
@@ -677,11 +677,7 @@ function ReaderStatistics:getStatisticEnabledMenuItem()
                 self:initData()
                 self.pages_stats = {}
                 self.start_current_period = TimeVal:now().sec
-                if self.document.info.has_pages then
-                    self.curr_page = self.ui.paging.current_page
-                else
-                    self.curr_page = self.document:getCurrentPage()
-                end
+                self.curr_page  = self.ui:getCurrentPage()
```
Comment (Frenzie) [inline plugins/statistics.koplugin/main.lua]:
```suggestion
                self.curr_page = self.ui:getCurrentPage()
```
